### PR TITLE
Upgrade helm-operator to Helm v3

### PR DESCRIPTION
**Pull request by joelanford**
**Description of the change:**
This PR bumps Helm to v3.0.0, adds the helm-2to3 plugin to auto migrate existing CRs, and refactors relevant portions of the helm operator code due to the restructuring between helm 2 and helm 3

**Motivation for the change:**
To support Helm 3 in Operator SDK

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -1,124 +1,119 @@
...
 	github.com/mattn/go-isatty v0.0.8
 	github.com/mitchellh/go-homedir v1.1.0
-	github.com/mitchellh/go-wordwrap v1.0.0 // indirect
 	github.com/mitchellh/mapstructure v1.1.2
-	github.com/mxk/go-flowrate v0.0.0-20140419014527-cca7078d478f // indirect
-	github.com/opencontainers/go-digest v1.0.0-rc1 // indirect
-	github.com/operator-framework/operator-lifecycle-manager v0.0.0-20190128024246-5eb7ae5bdb7a
+
+	// TODO: Can OLM use valid go modules version numbers?
+	github.com/operator-framework/operator-lifecycle-manager v0.0.0-20190926160646-a61144936680
```
Comment (njhale) [inline go.mod]: We've been wanting to, but some openshift automation added a [`v3.11` tag](https://github.com/operator-framework/operator-lifecycle-manager/tree/v3.11.0) to our repository, which would force us to start at a greater semver for minimal version selection to skip it. IIRC, this happened right as we started to use modules (go 1.11 timeframe). We should revisit this soon.

Comment (joelanford) [reply go.mod]: Oh no! That's a bummer. I don't envy you. However, since there's no `go.mod` file on the commit tagged with `v3.11.0`, you might not be totally hosed. See https://github.com/golang/go/wiki/Modules#semantic-import-versioning, specifically item 2 in the "Three Transitional Exceptions" section.

It might be worth trying to add a valid go module semver tag in a throwaway fork at `v0.12.0` to see whether you can import at that version without getting bumped "up" to `v3.11.0`. I'm pretty sure that will work.

Also since `v3.11.0` is not a go module (since it is lacking a `go.mod` file), you may even be able to delete the `v3.11.0` tag with little consequences. Don't quote me on that, but that's something else that might be worth trying in a throwaway fork.

File: test/test-framework/go.mod [hunk 2]
```diff
@@ -36,4 +36,6 @@ replace (
 	k8s.io/sample-apiserver => k8s.io/sample-apiserver v0.0.0-20191016112829-06bb3c9d77c9
 )
 
+replace github.com/docker/docker => github.com/moby/moby v0.7.3-0.20190826074503-38ab9da00309 // Required by Helm
```
Comment (camilamacedo86) [inline test/test-framework/go.mod]: The test-framework is a mock data of a Go project. So, why is required to add here and change its modules for the helm?

File: pkg/helm/release/manager_factory.go [hunk 3]
```diff
@@ -182,15 +252,29 @@ func getReleaseName(storageBackend *storage.Storage, crChartName string, cr *uns
...
+	if err != nil {
+		if notFoundErr(err) {
+			return nil, false, nil
+		}
+		return nil, false, err
+	}
+	return releaseHistory, len(releaseHistory) > 0, nil
+}
+
+func releaseHistoryV3(storageBackend *v3storage.Storage, releaseName string) ([]*helmreleasev3.Release, bool, error) {
```
Comment (camilamacedo86) [inline pkg/helm/release/manager_factory.go]: Why is required have releaseHistoryV2 and releaseHistoryV3? Will we keep the compatibility with old versions?
Should it still work on with helms v2?

File: internal/scaffold/go_mod.go [hunk 4]
```diff
@@ -68,6 +68,8 @@ replace (
 	k8s.io/metrics => k8s.io/metrics v0.0.0-20191016113814-3b1a734dba6e
 	k8s.io/sample-apiserver => k8s.io/sample-apiserver v0.0.0-20191016112829-06bb3c9d77c9
 )
+
+replace github.com/docker/docker => github.com/moby/moby v0.7.3-0.20190826074503-38ab9da00309 // Required by Helm
 `
```
Comment (camilamacedo86) [inline internal/scaffold/go_mod.go]: The same here. Why add the mod in a no helm project?

File: internal/scaffold/ansible/go_mod.go [hunk 5]
```diff
@@ -69,6 +69,8 @@ replace (
 	k8s.io/metrics => k8s.io/metrics v0.0.0-20191016113814-3b1a734dba6e
 	k8s.io/sample-apiserver => k8s.io/sample-apiserver v0.0.0-20191016112829-06bb3c9d77c9
 )
+
+replace github.com/docker/docker => github.com/moby/moby v0.7.3-0.20190826074503-38ab9da00309 // Required by Helm
 `
```
Comment (camilamacedo86) [inline internal/scaffold/ansible/go_mod.go]: same

File: test/test-framework/go.mod (hunk 2 above)
Comment (joelanford) [reply test/test-framework/go.mod]: Go project's depend on `github/operator-framework/operator-sdk`. Fro any projects resolving that dependency, this `replace` is required for resolution to succeed.

Here's what `go mod tidy` looks like in the test framework without that replace:
```console
$ go mod tidy
go: github.com/operator-framework/operator-sdk@v0.0.0 requires
	helm.sh/helm/v3@v3.0.0 requires
	github.com/deislabs/oras@v0.7.0 requires
	github.com/docker/docker@v0.0.0-00010101000000-000000000000: invalid version: unknown revision 000000000000
```

File: pkg/helm/release/manager_factory.go (hunk 3 above)
Comment (joelanford) [reply pkg/helm/release/manager_factory.go]: This is necessary to handle automatically migrating existing Helm v2 releases to Helm v3. The idea here is that a user with existing CRs and an existing Helm v2-based operator can upgrade to the new Helm 3-based operator and the new operator will handle the transition seamlessly.

File: internal/scaffold/go_mod.go (hunk 4 above)
Comment (joelanford) [reply internal/scaffold/go_mod.go]: See https://github.com/operator-framework/operator-sdk/pull/2080#discussion_r351479087

File: internal/scaffold/ansible/go_mod.go (hunk 5 above)
Comment (joelanford) [reply internal/scaffold/ansible/go_mod.go]: See https://github.com/operator-framework/operator-sdk/pull/2080#discussion_r351479087

File: internal/scaffold/helm/role.go [hunk 6]
```diff
@@ -186,21 +185,34 @@ func getServerVersionAndResources(dc roleDiscoveryInterface) (*version.Info, []*
...
+	install.Replace = true
+	install.ClientOnly = true
+	rel, err := install.Run(c, nil)
 	if err != nil {
 		return nil, fmt.Errorf("failed to render chart templates: %s", err)
 	}
-	return tiller.SortByKind(manifest.SplitManifests(renderedTemplates)), nil
+	_, manifests, err := releaseutil.SortManifests(releaseutil.SplitManifests(rel.Manifest), chartutil.DefaultVersionSet, releaseutil.InstallOrder)
+	return manifests, err
+	/*
```
Comment (camilamacedo86) [inline internal/scaffold/helm/role.go]: Should not it be removed?

File: internal/scaffold/helm/role.go [hunk 7]
```diff
@@ -186,21 +185,34 @@ func getServerVersionAndResources(dc roleDiscoveryInterface) (*version.Info, []*
...
+			},
+			KubeVersion: v,
+		}
+
+		renderedTemplates, err := renderutil.Render(c, &chart.Config{}, renderOpts)
+		if err != nil {
+			return nil, fmt.Errorf("failed to render chart templates: %s", err)
+		}
+		return tiller.SortByKind(releaseutil.SplitManifests(renderedTemplates)), nil
+	*/
```
Comment (camilamacedo86) [inline internal/scaffold/helm/role.go]: Should not it be removed? Why keep this code commented?

**Comment by jmccormick2001** [comment]
it could be my environment, but I built sdk for this PR, ran the user guide steps and am
seeing the following:

the CR shows size of 2....

    replicaCount: 2
    resources: {}
    securityContext: {}
    service:
      port: 8080
      type: ClusterIP
    serviceAccount:
      create: true
      name: null
    tolerations: []
kind: List
metadata:
  resourceVersion: ""
  selfLink: ""

but only a single pod and deployment is running....
jeffmc@doppio114 ~/projects/helmtest/nginx-operator> kubectl get pods
NAME                              READY   STATUS    RESTARTS   AGE
nginx-operator-6c7c94494b-9nc6m   1/1     Running   0          90s
jeffmc@doppio114 ~/projects/helmtest/nginx-operator> kubectl get deploy
NAME             READY   UP-TO-DATE   AVAILABLE   AGE
nginx-operator   1/1     1            1           119s

I'm seeing this in the operator log:
{"level":"error","ts":1575416482.6321547,"logger":"controller-runtime.controller","msg":"Reconciler error","controller":"nginx-controller","request":"default/example-nginx","error":"failed to load chart dir: apiVersion 'v2' is not valid. The value must be \"v1\"","stacktrace":"github.com/go-logr/zapr.


here is my test env:
jeffmc@doppio114 ~/projects/nginx-operator> kubectl version
Client Version: version.Info{Major:"1", Minor:"16", GitVersion:"v1.16.3", GitCommit:"b3cbbae08ec52a7fc73d334838e18d17e8512749", GitTreeState:"clean", BuildDate:"2019-11-13T11:23:11Z", GoVersion:"go1.12.12", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"16", GitVersion:"v1.16.3", GitCommit:"b3cbbae08ec52a7fc73d334838e18d17e8512749", GitTreeState:"clean", BuildDate:"2019-11-13T11:13:49Z", GoVersion:"go1.12.12", Compiler:"gc", Platform:"linux/amd64"}
jeffmc@doppio114 ~/projects/nginx-operator> go version
go version go1.13.1 linux/amd64

**Comment by joelanford** [comment]
@jmccormick2001 Yeah that happens because the `:master` base image in the project you created is still using the old code. You'll need to manually build the base image (`make image-build-helm`) and then update your Dockerfile to use that one.

**Comment by camilamacedo86** [comment]
Hi @jmccormick2001,

I created the image `cmacedo/helm-operator:helm-v3` please, feel free to use it to test.

**Comment by joelanford** [comment]
@camilamacedo86 Seems like the executable in your base image is built for an incorrect OS or architecture. Did you use `make image-build-helm`?

**Comment by camilamacedo86** [comment]
Yes, I used by I am running it from the Mac. So, I am approving this one since the test is passing and checking it and then, I will track an issue for we solve it in order to make the image builds correctly in a Mac OS as well.

File: cmd/operator-sdk/new/cmd.go [hunk 8]
```diff
@@ -308,8 +309,13 @@ func doHelmScaffold() error {
 		return fmt.Errorf("failed to create helm chart: %s", err)
 	}
 
-	valuesPath := filepath.Join("<project_dir>", helm.HelmChartsDir, chart.GetMetadata().GetName(), "values.yaml")
-	crSpec := fmt.Sprintf("# Default values copied from %s\n\n%s", valuesPath, chart.GetValues().GetRaw())
+	valuesPath := filepath.Join("<project_dir>", helm.HelmChartsDir, chart.Name(), "values.yaml")
+
+	rawValues, err := yaml.Marshal(chart.Values)
+	if err != nil {
+		return fmt.Errorf("failed to get raw chart values: %s", err)
```
Comment (estroz) [inline cmd/operator-sdk/new/cmd.go]:
```suggestion
		return fmt.Errorf("failed to get raw chart values: %v", err)
```
Not strictly necessary but at least it gives us the option of unwrapping in the caller, and aligns with other `fmt.Errorf`'s in the codebase.

File: doc/helm/dev/developer_guide.md [hunk 9]
```diff
@@ -118,36 +162,44 @@ spec:
           resources:
             {}
 
-
----
-# Source: foo/templates/ingress.yaml
-
+NOTES:
```
Comment (estroz) [inline doc/helm/dev/developer_guide.md]: This block should be outside of the YAML example correct?

File: internal/scaffold/helm/chart.go [hunk 10]
```diff
@@ -262,30 +249,25 @@ func createChartFromRemote(destDir string, opts CreateChartOptions) (*chart.Char
 		// One of Helm's error messages directs users to run `helm init`, which
 		// installs tiller in a remote cluster. Since that's unnecessary and
```
Comment (estroz) [inline internal/scaffold/helm/chart.go]: Is this bit on tiller still relevant?

File: internal/scaffold/helm/chart_test.go [hunk 11]
```diff
@@ -100,12 +66,12 @@ func TestCreateChart(t *testing.T) {
...
 		{
 			name:          "repo without helm chart",
-			helmChartRepo: repoURL,
+			helmChartRepo: srv.URL(),
 			expectErr:     true,
 		},
 		{
 			name:             "non-existent version",
-			helmChart:        latest.repoAndName,
+			helmChart:        "test/" + chartName,
```
Comment (estroz) [inline internal/scaffold/helm/chart_test.go]: If these are paths, they should be created using `filepath.Join`.

File: pkg/helm/release/manager.go [hunk 12]
```diff
@@ -194,42 +145,25 @@ func (m manager) getDeployedRelease() (*rpb.Release, error) {
...
-		Namespace: namespace,
-		Name:      name,
-		Chart:     chart,
-		Values:    config,
-	}
-
-	releaseResponse, err := tiller.InstallRelease(ctx, installReq)
+	installedRelease, err := install.Run(m.chart, m.values)
 	if err != nil {
 		// Workaround for helm/helm#3338
```
Comment (estroz) [inline pkg/helm/release/manager.go]: Is this comment and related code still relevant? The issue is closed.

Ditto below.

File: pkg/helm/release/manager_factory.go [hunk 13]
```diff
@@ -63,68 +67,133 @@ func (f managerFactory) NewManager(cr *unstructured.Unstructured) (Manager, erro
...
-	storageBackend := storage.Init(driver.NewSecrets(clientv1.Secrets(cr.GetNamespace())))
-	tillerKubeClient, err := client.NewFromManager(f.mgr)
-	if err != nil {
-		return nil, fmt.Errorf("failed to get client from manager: %s", err)
+
+	v2StorageBackend := v2storage.Init(v2driver.NewSecrets(clientv1.Secrets(cr.GetNamespace())))
+	v3StorageBackend := v3storage.Init(v3driver.NewSecrets(clientv1.Secrets(cr.GetNamespace())))
+
+	if err := convertV2ToV3(v2StorageBackend, v3StorageBackend, cr); err != nil {
+		return nil, fmt.Errorf("failed to convert releases from v2 to v3: %s", err)
```
Comment (estroz) [inline pkg/helm/release/manager_factory.go]: Since this method is exposed, formatted errors should be wrapped.

File: pkg/helm/release/manager_factory.go [hunk 14]
```diff
@@ -63,68 +67,133 @@ func (f managerFactory) NewManager(cr *unstructured.Unstructured) (Manager, erro
...
+	historyV2, existsV2, err := releaseHistoryV2(v2StorageBackend, releaseName)
 	if err != nil {
-		return nil, err
+		return err
 	}
-	cs, err := clientset.NewForConfig(kubeconfig)
-	if err != nil {
-		return nil, err
+	if existsV2 {
+		if err := convertHistoryToV3(historyV2, v2StorageBackend, v3StorageBackend); err != nil {
```
Comment (estroz) [inline pkg/helm/release/manager_factory.go]:
```suggestion
		return convertHistoryToV3(historyV2, v2StorageBackend, v3StorageBackend)
```
And modify below code.

File: pkg/helm/release/manager_factory.go [hunk 15]
```diff
@@ -63,68 +67,133 @@ func (f managerFactory) NewManager(cr *unstructured.Unstructured) (Manager, erro
...
+	legacyHistoryV2, legacyV2Exists, err := releaseHistoryV2(v2StorageBackend, legacyName)
+	if err != nil {
+		return err
 	}
-	env := &environment.Environment{
-		EngineYard: ey,
-		Releases:   storageBackend,
-		KubeClient: tillerKubeClient,
+	if legacyV2Exists {
+		if err := convertHistoryToV3(legacyHistoryV2, v2StorageBackend, v3StorageBackend); err != nil {
```
Comment (estroz) [inline pkg/helm/release/manager_factory.go]:
```suggestion
		return convertHistoryToV3(legacyHistoryV2, v2StorageBackend, v3StorageBackend)
```

File: pkg/helm/release/manager_factory.go [hunk 16]
```diff
@@ -63,68 +67,133 @@ func (f managerFactory) NewManager(cr *unstructured.Unstructured) (Manager, erro
...
-	e := engine.NewOwnerRefEngine(baseEngine, restMapper, ownerRefs)
-	var ey environment.EngineYard = map[string]environment.Engine{
-		environment.GoTplEngine: e,
+	return resourceList, nil
+}
+
+func convertV2ToV3(v2StorageBackend *v2storage.Storage, v3StorageBackend *v3storage.Storage, cr *unstructured.Unstructured) error {
+	// If a v2 release with the legacy name exists, convert it to v3.
+	legacyName := getLegacyName(cr)
+	legacyHistoryV2, legacyV2Exists, err := releaseHistoryV2(v2StorageBackend, legacyName)
```
Comment (estroz) [inline pkg/helm/release/manager_factory.go]: Nit: some variable/function names have `*V2` or `*V3` name formats, while others have `v2*` or `v3*`; it would be nice to give all names the same format.

File: pkg/helm/release/manager_factory.go [hunk 17]
```diff
@@ -63,68 +67,133 @@ func (f managerFactory) NewManager(cr *unstructured.Unstructured) (Manager, erro
...
+	return nil
+}
+
+func convertHistoryToV3(history []*helmreleasev2.Release, v2StorageBackend *v2storage.Storage, v3StorageBackend *v3storage.Storage) error {
+	for _, v2Rel := range history {
+		v3Rel, err := helm2to3.CreateRelease(v2Rel)
+		if err != nil {
+			return fmt.Errorf("generate v3 release: %s", err)
+		}
+		if err := v3StorageBackend.Create(v3Rel); err != nil {
```
Comment (estroz) [inline pkg/helm/release/manager_factory.go]: Is creating the new release before deleting the old ok, i.e. no conflicts? Different backends are used so I doubt there will be any issues, just checking.

**Comment by jmccormick2001** [comment]
seems to test out ok on OCP 4.2, I deployed using the master branch (helm v2), then deleted the nginx-operator, then built a new nginx-operator against this PR code, redeployed the nginx-operator, then updated the CR to add a new nginx pod, seems to work.

File: cmd/operator-sdk/new/cmd.go (hunk 8 above)
Comment (joelanford) [reply cmd/operator-sdk/new/cmd.go]: To do unwrapping, I'm pretty sure we need `%w` (right?), but yeah I agree we should just move all `fmt.Errorf()` and `errors.Wrapf()` calls to this. I'll update this one and any others I stumble across in the Helm code.

File: doc/helm/dev/developer_guide.md (hunk 9 above)
Comment (joelanford) [reply doc/helm/dev/developer_guide.md]: No, that's part of the output of `helm install --dry-run`

File: internal/scaffold/helm/chart.go (hunk 10 above)
Comment (joelanford) [reply internal/scaffold/helm/chart.go]: Nope, removing.

File: internal/scaffold/helm/chart_test.go (hunk 11 above)
Comment (joelanford) [reply internal/scaffold/helm/chart_test.go]: These aren't paths. They're `<repo>/<chartName>` strings for remote paths. But there are some other paths that need `filepath.Join`

File: pkg/helm/release/manager.go (hunk 12 above)
Comment (joelanford) [reply pkg/helm/release/manager.go]: I tried using the new `Atomic` config item, but that changed the semantics in a way I didn't expect. IIRC, it also forces `Wait`, which cascaded into some other issues. I think we should revisit this, but maybe in a follow-up?

File: pkg/helm/release/manager_factory.go (hunk 16 above)
Comment (joelanford) [reply pkg/helm/release/manager_factory.go]: I think I've got them all covered. Let me know if you see anymore I missed.

File: pkg/helm/release/manager_factory.go (hunk 17 above)
Comment (joelanford) [reply pkg/helm/release/manager_factory.go]: Yes. In fact that explicitly desired so that we are guaranteed that the new release has been created before we delete the old. This ensures no data loss. If we did the reverse, it would be possible for a delete to succeed and the subsequent create to fail.

File: cmd/operator-sdk/new/cmd.go (hunk 8 above)
Comment (joelanford) [reply cmd/operator-sdk/new/cmd.go]: I also created #2308 to track the follow-on work.

Comment (estroz) [reply cmd/operator-sdk/new/cmd.go]: If `err`'s underlying type implements `Unwrap`, `%v` will still allow unwrapping. Agreed though, all errors should be wrapped where it makes sense.
